packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
+    // of active proposal IDs, requires keys to be non-zero.
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
       exchanger: msg.sender,
       stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
       sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
       buyAmount: buyAmount,
       approvalTimestamp: 0, // initial value when not approved yet
-      state: ExchangeState.Proposed,
+      state: ExchangeProposalState.Proposed,
       sellCelo: sellCelo
     });
-    exchangeProposalCount = exchangeProposalCount.add(1);
+    // Push it into the array of active proposals.
+    activeProposalIds.push(exchangeProposalCount);
     // Even if stable tokens are being sold, the sellAmount emitted is the "value."
     emit ExchangeProposalCreated(
-      proposalId,
+      exchangeProposalCount,
       msg.sender,
       stableToken,
       sellAmount,
       buyAmount,
       sellCelo
     );
+    return exchangeProposalCount;
+  }
 
-    return proposalId;
+  /**
+   * @notice Approves an existing exchange proposal.
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Set the time the approval occurred and change the state.
+    proposal.approvalTimestamp = block.timestamp;
+    proposal.state = ExchangeProposalState.Approved;
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
+      "Sender cannot cancel the exchange proposal"
+    );
+    // Mark the proposal as cancelled. Do so prior to refunding as a measure against reentrancy.
+    proposal.state = ExchangeProposalState.Cancelled;
+    // Remove the proposal from the active list.
+    activeProposalIds.remove(proposalId);
```

Comment:
just double checking, IIUC this list has an index of elements, so this call is `O(1)` instead of a `O(N)` lookup, right?